### PR TITLE
[SYCL][Graph] Add support for fill and memset nodes in graphs

### PR DESCRIPTION
**Pull request by Bensuo**
- Adds support for fill and memset nodes in graphs.
- Supported on Level Zero only for now.
- Adds E2E and unit tests for these new node types.
- Minor modifications due to renaming of some UR functions.

### CODE REVIEW

**Comment by al42and** [comment]
May I ask when this change and the associated UR PR are expected to be merged?

**Comment by al42and** [comment]
BTW, is it the expected behavior that

```cpp
q.submit((sycl::handler & cgh) { cgh.memset(ptr_, 0, size_); });
```

does not work without this PR, but

```cpp
q.memset(ptr_, 0, size_);
```

works fine in a graph on LevelZero devices with the current `sycl` branch (at least in a very simple test)?

File: sycl/plugins/opencl/pi_opencl.cpp
Comment (mfrancepillois) [reply sycl/plugins/opencl/pi_opencl.cpp]: These stubs has been moved to UR

**Comment by EwanC** [comment]
I've been struggling to understand the discrepancy between the `code_formatter` CI job and pre-commit windows/linux clang-format jobs, as they expect different formats for the pi.h changes.

It looks like we are transitioning to a new formatter https://github.com/intel/llvm/pull/12199#issuecomment-1874228112

Gone with the old style of formatting, as otherwise the windows/linux CI jobs are blocked on the early fail.

**Review by mfrancepillois — APPROVED** [review]

**Comment by aelovikov-intel** [comment]
> I've been struggling to understand the discrepancy between the `code_formatter` CI job and pre-commit windows/linux clang-format jobs, as they expect different formats for the pi.h changes.
> 
> It looks like we are transitioning to a new formatter [#12199 (comment)](https://github.com/intel/llvm/pull/12199#issuecomment-1874228112)
> 
> Gone with the old style of formatting, as otherwise the windows/linux CI jobs are blocked on the early fail. @aelovikov-intel is this the best path forward? I could also disable clang-format for these new lines with `// clang-format off/on` and then we remove that guard once the CI formatter switch is done

I'm removing the old formatter in https://github.com/intel/llvm/pull/12308. If the current version of the PR will be approved, then failing new formatter won't be a blocker for the merge. Otherwise, once #12308 is merged, any subsequent updates to this PR would have to be updated to pass the new one.

File: sycl/doc/design/CommandGraph.md
Comment (aelovikov-intel) [inline sycl/doc/design/CommandGraph.md:40]: Formatting looks wrong here and on line 47.

File: sycl/include/sycl/detail/pi.h
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/pi.h:2469]:
```suggestion
/// \param buffer is the location to fill the data.
```

File: sycl/test-e2e/Graph/Explicit/buffer_fill.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/Graph/Explicit/buffer_fill.cpp:4]:
```suggestion
// RUN: %if level_zero %{env UR_L0_LEAKS_DEBUG=1 %{run} %t.out 2>&1 | FileCheck --implicit-check-not=LEAK %s %}
```

and drop line 6. Same in other file(s).

File: sycl/test-e2e/Graph/Inputs/buffer_fill.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/Graph/Inputs/buffer_fill.cpp:11]: Unrelated to this PR, but I think the support should be printed as part of `sycl-ls` and then handled via `REQUIRES` to properly mark skipped tests.

File: sycl/test-e2e/Graph/Inputs/buffer_fill.cpp
Comment (aelovikov-intel) [inline sycl/test-e2e/Graph/Inputs/buffer_fill.cpp:7]: I assume https://github.com/intel/llvm/pull/12279 would need some changes.

File: sycl/unittests/Extensions/CommandGraph.cpp
Comment (aelovikov-intel) [inline sycl/unittests/Extensions/CommandGraph.cpp:2182]: FYI, AFAIK, we're not running unittests in CI. We have an internal bug report for this, but I don't know anyone actively working on it.

File: sycl/test-e2e/Graph/Explicit/buffer_fill.cpp
Comment (aelovikov-intel) [reply sycl/test-e2e/Graph/Explicit/buffer_fill.cpp:4]: It seems that https://github.com/intel/llvm/pull/12279 will update existing tests to the spelling I've suggested. Up to the graph devs on what the format for newly added tests should be.

**Review by kbenzie — APPROVED** [review]
UR LGTM

File: sycl/test-e2e/Graph/Explicit/buffer_fill.cpp
Comment (EwanC) [reply sycl/test-e2e/Graph/Explicit/buffer_fill.cpp]: Updated this PR to use `--implicit-check-not=LEAK`, since that should be the format going forward.

File: sycl/unittests/Extensions/CommandGraph.cpp
Comment (EwanC) [reply sycl/unittests/Extensions/CommandGraph.cpp:2182]: That's good to know, I wasn't aware of that.

Comment (aelovikov-intel) [reply sycl/unittests/Extensions/CommandGraph.cpp:2182]: One can run `check-sycl-unittests` locally if they wish :)

File: sycl/test-e2e/Graph/Inputs/buffer_fill.cpp
Comment (EwanC) [reply sycl/test-e2e/Graph/Inputs/buffer_fill.cpp:11]: That's a good idea we hadn't considered, I've created an internal ticket for our team to do this as a follow-up standalone PR for all tests.

**Comment by EwanC** [comment]
Thanks for the review @aelovikov-intel, I think I've addressed all your comments if you could take another look.

**Review by aelovikov-intel — APPROVED** [review]

**Comment by EwanC** [comment]
@intel/llvm-gatekeepers Can you merge this please

**Comment by EwanC** [comment]
This PR broke post-commit CI due to an unused variable, I'll look into it.

File: sycl/unittests/Extensions/CommandGraph.cpp
Comment (aelovikov-intel) [inline sycl/unittests/Extensions/CommandGraph.cpp:2209]: Does anybody know why are we checking `std::shared_ptr::operator==` here? Should the comment be that nodes must be different? Or check `operator==` of something else?

Comment (Bensuo) [reply sycl/unittests/Extensions/CommandGraph.cpp:2209]: Hmm yes I think this is probably a mistake and should probably be checking equality of the `node` objects not the impls if anything, since the `node` class does have an `operator==` defined. Though it's just going to compare the impl pointers anyway.

**Pull request merged**
